### ISSUE TITLE
Implement IMemoryOwner<T> on Array

### ISSUE THREAD

**Issue by davidfowl** [dotnet/coreclr#22873]
I've been thinking about clean ways to pass around the `T[]` returned from an `ArrayPool<T>` without having to pass around a separate owner object that represents the memory pool and the array itself. Instead it would be cleaner if we could pass a `T[]` around as an `IMemoryOwner<T>` without having to allocate an extra object. On top of that, we'd need to figure out how we could make the `IMemoryOwner<T>` work for array pool backed arrays (we'd need to store that bit of information somewhere in the array layout) so that disposing the owner returns the memory back to the pool. 

This would also simplify the `MemoryPool<byte>.Shared` implementation and make it non-allocating.

cc @jkotas @stephentoub

**Comment by stephentoub** [comment]
I do not think this is a good thing to do. It would make every array disposable, which is terribly confusing. It would put pooling in the face of every .NET developer. It would make use-after-free errors really likely. It would add non-trivial additional costs to what is a critically core type. It would tie that type to a specific pooling design. And probably a bunch of other concerns.
